Loader: Intro, `fabric.mod.json`, Dependency Overrides (#557)
FabricMC/fabric-docs: develop/loader/index.md

The mod ID identifies the mod, and any two mods with the same ID are considered to be the same mod. Only one version of a mod may be loaded at a time.

A mod may also declare which other mods it depends on or conflicts with. Fabric Loader will attempt to satisfy dependencies and load the appropriate versions of mods; if it cannot it will fail to launch the game.

Fabric Loader makes all mods equally capable of modifying the game. As an example, anything Fabric API does can be done by any other mod.

Mods are loaded both from the classpath and from the `mods` directory. This directory can be changed with the `fabric.modsFolder` system property.

## Nested JARs {#nested-jars}

::: info

Using Fabric Loom's `include` option will automatically handle nesting the JAR, including the generation of a `fabric.mod.json` for non-mod JARs.

:::

Nested JARs allow a mod to provide its own dependencies, so Fabric Loader can find the best version while attempting to satisfy dependencies, instead of requiring them to be installed separately.

> Nested jars act like any other mod, with their own metadata file, but are contained within the parent JAR. Note that nested mods may themselves embed other children in the same way.

Reviewer: ```suggestion
Nested JARs act like any other mod, with their own metadata file, but are contained within the parent JAR. Note that nested mods may themselves embed other children in the same way.
```

Nested JARs are extracted to the disk when the game is run. Paths to nested JARs must be declared relative to their containing JAR's root.

## Entrypoints {#entrypoints}

> Fabric Loader has an [entrypoint](../getting-started/project-structure#entrypoints) system, which is used by mods to expose parts of the code for usage by Fabric Loader or other mods. Fabric Loader uses it for mod initialization. Initializers are loaded and called early during the game's initialization which allows a mod to run some code to make its modifications. These entrypoints are typically used to bootstrap mods by registering registry objects, event listeners and other callbacks for doing things later.

Reviewer: ```suggestion
Fabric Loader adopts an [entrypoint](../getting-started/project-structure#entrypoints)-based system, which is used to expose certain code to Fabric Loader for initialization, or to other mods.

Initializers are loaded and called early during the game's boot, and this allows a mod to make its modifications with code. These entrypoints are typically used to bootstrap mods by registering registry objects, event listeners, and other callbacks for later actions.
```

## Mixin {#mixin}

> Mixin allows mods to transform Minecraft classes and even mod classes, and is the only method of class transformation that Fabric Loader officially supports. A mod can declare its own mixin configuration which enables the use of Mixin.

Reviewer: ```suggestion
Mixins allow mods to transform Minecraft classes, and even other mods' classes, and they are the only type of class transformation that Fabric Loader officially supports. A mod can declare its own mixin configuration, which enables the use of Mixin.
```

based on https://discord.com/channels/507304429255393322/1208846408552030238/1447873449748987985

> Mixin was not originally made for Fabric, so Fabric Loader uses a modified version of Mixin. However, the [documentation](https://maven.fabricmc.net/docs/sponge-mixin-0.17.1+mixin.0.8.7/) of the upstream version is still mostly valid. Fabric's modifications include allowing all default injection points inside constructors, optimizing out unused callback infos, providing fixes for backwards compatibility, fixing static shadows, allowing injectors in interfaces, and more.

Reviewer: ```suggestion
Fabric Loader uses a modified fork of the original Sponge Mixin. However, the [upstream Mixin wiki](https://github.com/SpongePowered/Mixin/wiki) is still mostly valid.

Among Fabric's modifications there are: allowing all default injection points inside constructors, optimizing out unused callback infos, providing fixes for backwards compatibility, fixing static shadows, allowing injectors in interfaces, and more.
```

At least I think that's the intended link?

<!-- Referenced comments from LlamaLad7 on Fabric Mixin. https://discord.com/channels/507304429255393322/566418023372816394/1002211121903706162 -->

> 

Reviewer: ```suggestion
```

LlamaLad7 is already added as an author, and the linked message is already completely included in the paragraph above.

## Mappings {#mappings}

::: info

Mappings are only relevant when using Fabric Loader on obfuscated games, including pre-26.1 versions of Minecraft.

:::

Fabric Loader provides the `MappingResolver` API to determine names of classes, fields and methods with respect to the different environments that mods may be loaded in. This can be used to support reflection in any environment provided Fabric Loader has access to mappings to resolve the name.

```java
FabricLoader.getInstance().getMappingResolver().mapClassName("intermediary", "net.minecraft.class_5421") // Resolves to `RecipeBookType` on named versions of 1.21.11
```

> When launched in a non-development environment on an obfuscated game, Fabric Loader will [remap](../porting/mappings/index#mappings) the game jar(s) to intermediary names. Mods designed for obfuscated games are expected to be mapped to intermediary, which will be compatible with this environment. The remapped jars are cached and saved in `${gameDir}/.fabric/remappedJars/${minecraftVersion}` for re-use across launches.

Reviewer: not too keen on adding this section since it's not relevant to 26.1 anymore. is `MappingResolver` still planned to be supported?

Author: There's no plan to remove it, and does still make sense for the few people running Modern Yarn on 26.1. Loader is version independent, and these docs reflect that.
